plugin-api/src/main/java/org/sonar/api/server/http/HttpRequest.java

```diff
@@ -0,0 +1,107 @@
+ */
+package org.sonar.api.server.http;
+
+import java.util.Enumeration;
+
+/**
+ * Framework-agnostic definition of an HTTP request.
+ *
+ * @since 9.16
+ */
+public interface HttpRequest {
+
+  /**
+   * Returns the port number to which the request was sent.
+   */
+  int getServerPort();
+
+  /**
+   * Returns a boolean indicating whether this request was made using a secure channel, such as HTTPS.
+   */
+  boolean isSecure();
+
+  /**
+   * Returns the name of the scheme used to make this request, for example, http, https, or ftp.
+   */
+  String getScheme();
+
+  /**
+   * Returns the host name of the server to which the request was sent.
+   */
+  String getServerName();
+
+  /**
+   * Returns the URL the client used to make the request. The returned URL contains a protocol, server name, port number, and server path,
+   * but it does not include query string parameters.
+   */
+  String getRequestURL();
+
+  /**
+   * Returns the part of this request's URL from the protocol name up to the query string in the first line of the HTTP request.
+   */
+  String getRequestURI();
+
+  /**
+   * Returns the query string that is contained in the request URL after the path. This method returns null if the URL does not have a
+   * query string.
+   */
+  String getQueryString();
+
+  /**
+   * Returns the portion of the request URI that indicates the context of the request. The context path always comes first
+   * in a request URI. The path starts with a "/" character but does not end with a "/" character. For servlets in the
+   * default (root) context, this method returns "". The container does not decode this string.
+   */
+  String getContextPath();
+
+  /**
+   * Returns the value of a request parameter as a String, or null if the parameter does not exist.
+   * You should only use this method when you are sure the parameter has only one value. If the parameter might have more than one value,
+   * use {@link #getParameterValues}.
+   */
+  String getParameter(String name);
+
+  /**
+   * Returns an array containing all the values the given request parameter has, or null if the parameter does not exist.
+   */
+  String[] getParameterValues(String name);
+
+  /**
+   * Returns the value of the specified request header as a String. If the request did not include a header of the specified name, this
+   * method returns null. If there are multiple headers with the same name, this method returns the first head in the request.
+   */
+  String getHeader(String name);
+
+  /**
+   * Returns all the values of the specified request header as an Enumeration of String objects.
+   */
+  Enumeration<String> getHeaders(String name);
+
+  /**
+   * Returns the name of the HTTP method with which this request was made, for example, GET, POST, or PUT.
+   */
+  String getMethod();
+
+  /**
+   * Returns the raw request object from the Servlet API that matches this request.
+   */
+  Object getRawRequest();
```

Comment:
This seems to defeat the purpose of having our own objects that represent HTTP requests and responses.

If the goal is for this to be used only by the products implementing the API, we should cast classes instead, since they are in control of the implementations at runtime.

For example, if a given version of SonarQube is creating filter contexts with  `HttpRequest`  of type `JakartaHttpRequest`, it's save for its filters to cast `HttpRequest` to `JakartaHttpRequest`.

Reply:
Hmm, so we wanted to return "raw" requests objects as we wanted to avoid creating more classes that mimic `javax.servlet.*` or `jakarta.servlet.*` classes, like for example `Cookies`. Though maybe indeed you are right as it defeats the purpose and with a little more effort and we could get rid of all classes instead of taking shortcuts.